### DISCUSSION TITLE
[MODDING] "NoteMods" or per-note modifiers.

### DISCUSSION THREAD

**Discussion by MrThatKid** [TeamRizu/OutFox#335] [03 - Feature Requests]
Video showing the proof-of-concept: [https://youtu.be/dapx_2RTESc](https://youtu.be/dapx_2RTESc)
This allows one to set a variety of modifiers on a single note, allowing for ever further modding potential.

There are two ways to add per-note mods currently.

Method 1: Directly in the simfile's notedata. (using dance's notedata layout as an example)

- Say you have a down note: `0100`
- This is a down note with dizzy: `01<dizzy>00`
- You can have multiple mods on a single note: `01<dizzy/drunk>00`
- Magnitudes are also a thing you can do (both positive and negative): `01<314 dizzy/-50.1234 drunk>00`
  - Like modstrings, you can put a % after the magnitude, but it's optional.

Method 2: Through lua.

- Player now has a AddNoteMod function, which allows for function chaining like the other Player functions.
- It takes the beat, column, mod name and the optional magnitude. (when not given, the default is 100%)

```lua
player:AddNoteMod(16, 1, "bumpy", 2)
```
This sets the note in the first column at beat 16 to have 200% bumpy on. If there is no note at that spot, nothing happens.
If the column is out of range, an error is raised.

Behavior:
- This is additive to the notefield modifiers (but is separate from the column-specific modifiers if there is both field and column variants)
  - For example, it adds onto MoveX (which is column-specific), but it can also add onto Bumpy (with column-specific bumpy being considered a separate thing).
- Mod names are case-insensitive, but are stored in lowercase.

Limitations:
- There is no easy way to clear a note's mods, as well as no way to apply them in batches (with a single function).
- Most of the sub-modifiers are missing, as well as a few other mods & boolean mods.
- Due to how these are stored, there is no way to do approach rates or tweening without doing something perframe.
  - It is unknown how setting a notemod perframe will impact performance.
- The editor will have to be altered to either not have the notemods show up or do something with the mods a note has.

If you've got any suggestions for what you'd want for this or what should change, please say something!
	- MTK

**Comment by Sudospective** [comment]
Changing the name to something like `ApplyNoteMod` might be good, considering `AddNoteMod` implies that it literally adds to the note modifier.

That being said, I'm pushing to a staging branch of my Kitsu template that allows you to insert and ease note-specific mods. Right now, there's no Mirin or Exschwasion function counterpart, but `InsertNoteMod` will let you insert one or more notetables: `{beat, column, percent, mod}` and apply them in batches. In theory, this will let you clear the note mods as well, but my clearmod code isn't quite there figuratively and literally.

Now, what about using note mods and note data together...? 🤔

**Reply by Sudospective** [reply]
https://www.youtube.com/watch?v=fuzIcjOU-n4 A quick and dirty video demo using a fabricated notedata table

**Reply by Sudospective** [reply]
Also, for anyone curious, yes, setting a notemod perframe in its current state will impact performance. A lot. I can't imagine someone doing this for every single note of an entire file, but it is possible, albeit with destroyed framerate.
